### ISSUE TITLE
exclusion constraint validation

### ISSUE THREAD

**Issue by jw1u1** [pgpartman/pg_partman#529]
Hi,
creating exclusion constraints as invalid requires manual interference to validate them.
Please consider to automate the task of constraint validation with minimal lock.
* Get a sharelock on the table
* create the constraint not valid
* update the system catalogue to validate the constraint

**Comment by keithf4** [comment]
Updating the system catalog directly like that is giving a false impression that all the data in that table is actually valid for that constraint. I likely will not be doing something like that in pg_partman.

**Comment by jw1u1** [comment]
That's what the sharelock is for...

**Comment by keithf4** [comment]
Sorry, not following you then. I read your suggestion as you're creating the constraint as not valid. This means it completely skips validating the existing data in the table. Then I assumed you meant you were directly editing the system catalog to mark the constraint as valid vs running VALIDATE CONSTRAINT. 

Is that no what you're suggesting?

**Comment by jw1u1** [comment]
Yes, in one transaction with a sharelock on the table, so concurrent writes are blocked.
You could even add a fast validation by checking min an max values of the column.

**Comment by keithf4** [comment]
Sorry but I'm not going to support editing system catalogs like this at this time. You're hacking the catalog to say something is valid in a manner that does not match what PG says it should mean. When a constraint is marked as valid, that should mean that all data in that constraint should have been validated to PostgreSQL's satisfaction.

**Comment by jw1u1** [comment]
Please find a solution to automate non-blocking exclusion Constraint creation.
The lack of this feature is a big disadvantage for large Databases with high availability demands.

**Comment by keithf4** [comment]
The method of directly hacking the catalog to make a constraint "valid" is one solution, but it is generally not recommended because you are breaking the contract that PG has made on what it considers a valid constraint. You may know what you've done, but unless you have it clearly documented for all users of your database, no one else will know that the constraint has been hacked. It is not a good long term solution. For it to truly be valid, all data in the table that it controls must be validated somehow. The method of doing the NOT VALID constraint first then running the validation later does allow for a lower level lock to be obtained when validating the existing data, so that is the only method currently available that I know of.

https://www.postgresql.org/docs/15/sql-altertable.html#SQL-ALTERTABLE-NOTES

This issue is honestly outside the scope of pg_partman at this time. This extension itself does not make exclusion constraints and they are not something that is exclusive to partitioning. This sort of request would be better made on the one of mailing lists (general or hackers) to see if there's some more generalized solution in core. 

https://www.postgresql.org/list/

**Comment by keithf4** [comment]
Closing this issue for now. Please feel free to reopen if you have any further questions
